### ISSUE TITLE
Enhancement: naming-convention - optional prefixes

### ISSUE THREAD

**Issue by roryabraham** [typescript-eslint/typescript-eslint#9554]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/naming-convention/#selector-options

### Description

I'd love to see an additional configuration called `optionalPrefix` and/or `optionalSuffix`, _or_ to allow the `prefix` and `suffix` configs to accept an empty string (which would be equivalent to making any of the listed prefixes optional).

### Fail

```TypeScript
const myCamelCaseVar = 4;
const my_snake_case_var = 5;
```

### Pass

```TypeScript
const myCamelCaseVar = 4;
const MyAllowedPrefix_myCamelCaseVar = 5;
```

### Additional Info

We have some use-cases where we want to prefix some variables with what's essentially a namespace, such as `private_` or `account_`. We don't want to generally allow snake case for variable names, but we want to set up a whitelist of allowed prefixes which may contain underscores like this.

Whether or not this is "niche" is up for debate, but I figured that this would be an easy enough change - just remove the additional validation that prohibits an empty string in `prefix`. This seems like it would make the rule more flexible/powerful, without much downside to users who don't need it.

**Comment by roryabraham** [comment]
Another alternate solution that might work well for us would be exporting the regexes used for the standard rules, such as `camelCase` or `UPPER_SNAKE_CASE`, such that we can create a custom "wrapper regex" to add the optional prefixes we want without having to re-implement all the regexes you've written and tested for this rule.

**Comment by roryabraham** [comment]
Yet another alternative would be, again, to export the matcher functions used internally for common naming conventions and allow `custom` to be a function instead of `{match, regex}`

**Comment by Josh-Cena** [comment]
Hi! Please see the note at the top of this doc page:

> This rule is feature frozen: it will no longer receive new features such as new options. It still will accept bug and documentation fixes for its existing area of features and to support new TypeScript versions.
>
> Stylistic rules that enforce naming and/or sorting conventions tend to grow incomprehensibly complex as increasingly obscure features are requested. This rule has reached the limit of what is reasonable for the typescript-eslint project to maintain. See [eslint-plugin: Feature freeze naming and sorting stylistic rules](https://github.com/typescript-eslint/typescript-eslint/issues/8792) for more information.

All feature requests are automatically rejected. Thanks for understanding.
